lib/Runtime/Debug/TTSnapObjects.cpp

```diff
@@ -2117,15 +2128,277 @@ namespace TTD
+            Js::ScriptContext* ctx = inflator->LookupScriptContext(snpObject->SnapType->ScriptContextLogId);
+
+            SnapGeneratorInfo* generatorInfo = SnapObjectGetAddtlInfoAs<SnapGeneratorInfo*, SnapObjectType::SnapGenerator>(snpObject);
+
+            Js::Var* argVals = nullptr;
+            if (generatorInfo->arguments_count > 0) {
+                argVals = RecyclerNewArray(ctx->GetRecycler(), Field(Js::Var), generatorInfo->arguments_count);
+
+                for (Js::RegSlot i = 0; i < generatorInfo->arguments_count; i++) {
+                    argVals[i] = inflator->InflateTTDVar(generatorInfo->arguments_values[i]);
+                }
+            }
+
+            Js::ScriptFunction*  scriptFunction = nullptr;// static_cast<Js::ScriptFunction*>((generatorInfo->scriptFunction != TTD_INVALID_PTR_ID) ? inflator->LookupObject(generatorInfo->scriptFunction) : nullptr);
+
+            Js::CallInfo callInfo;
+            memcpy(&callInfo, generatorInfo->arguments_callInfo, sizeof(Js::CallInfo));
+
+            Js::Arguments arguments(callInfo, argVals);
+
+            // TODO: BUGBUG - figure out how to determine what the prototype was.  Just use GetNull() for now
+            Js::RecyclableObject* prototype = ctx->GetLibrary()->GetNull();
+            //if (generatorInfo->generatorPrototype == 1) {
+            //    prototype = ctx->GetLibrary()->GetNull();
+            //}
+            //else if (generatorInfo->generatorPrototype == 2) {
+            //    prototype = ctx->GetLibrary()->CreateGeneratorConstructorPrototypeObject();
+            //}
+            //else {
+            //    //TTDAssert(false, "unexpected prototype found JavascriptGenerator");
+            //}
+
+            //Js::DynamicType* type = reinterpret_cast<Js::DynamicType*>((snpObject->SnapType->TypePtrId != TTD_INVALID_PTR_ID) ? inflator->LookupType(snpObject->SnapType->TypePtrId): nullptr);
+
+            return  ctx->GetLibrary()->CreateJavascriptGenerator_TTD(ctx, prototype, arguments, scriptFunction, static_cast<Js::JavascriptGenerator::GeneratorState>(generatorInfo->state));
+        }
+
+        void DoAddtlValueInstantiation_SnapGeneratorInfo(const SnapObject* snpObject, Js::RecyclableObject* obj, InflateMap* inflator) {
+            Js::ScriptContext* ctx = inflator->LookupScriptContext(snpObject->SnapType->ScriptContextLogId);
+            Js::JavascriptGenerator* generator = reinterpret_cast<Js::JavascriptGenerator*>(obj);
+
+            SnapGeneratorInfo* generatorInfo = SnapObjectGetAddtlInfoAs<SnapGeneratorInfo*, SnapObjectType::SnapGenerator>(snpObject);
+            Js::ScriptFunction* scriptFunction = reinterpret_cast<Js::ScriptFunction*>((generatorInfo->scriptFunction != TTD_INVALID_PTR_ID) ? inflator->LookupObject(generatorInfo->scriptFunction) : nullptr);
+            generator->SetScriptFunction(scriptFunction);
+
+            Js::FunctionBody* executeFunction = Js::JavascriptFunction::UnsafeFromVar(scriptFunction)->GetFunctionBody();
```

Comment:
Why is the `UnsafeFromVar` required here? Doesn't `Js::ScriptFunction` inherit from `Js::JavascriptFunction` and so `GetFunctionBody` should already be a member?

Reply:
And I just saw that `scriptFunction` can be `nullptr` if `generatorInfo->scriptFunction` is `TTD_INVALID`; is that a recoverable state, or should we failfast?

Reply:
fixed - removed `UnsafeFromVar` call & added an `AssertOrFailFast` call if scriptFunction is null.